service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CreateTableCopyJobsStep.java

```diff
@@ -0,0 +1,194 @@
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.api.services.bigquery.Bigquery;
+import com.google.api.services.bigquery.model.Job;
+import com.google.api.services.bigquery.model.JobConfiguration;
+import com.google.api.services.bigquery.model.JobConfigurationTableCopy;
+import com.google.api.services.bigquery.model.Table;
+import com.google.api.services.bigquery.model.TableList;
+import com.google.api.services.bigquery.model.TableList.Tables;
+import com.google.api.services.bigquery.model.TableReference;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateTableCopyJobsStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(CreateTableCopyJobsStep.class);
+  public static final Duration COPY_JOB_TIMEOUT = Duration.ofHours(12);
+  private final CrlService crlService;
+  private final WorkspaceService workspaceService;
+  private final ControlledBigQueryDatasetResource sourceDataset;
+
+  public CreateTableCopyJobsStep(
+      CrlService crlService,
+      WorkspaceService workspaceService,
+      ControlledBigQueryDatasetResource sourceDataset) {
+    this.crlService = crlService;
+    this.workspaceService = workspaceService;
+    this.sourceDataset = sourceDataset;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final FlightMap workingMap = flightContext.getWorkingMap();
+    // Gather inputs
+    final DatasetCloneInputs sourceInputs = getSourceInputs();
+    workingMap.put(ControlledResourceKeys.SOURCE_CLONE_INPUTS, sourceInputs);
+
+    final DatasetCloneInputs destinationInputs = getDestinationInputs(flightContext);
+    workingMap.put(ControlledResourceKeys.DESTINATION_CLONE_INPUTS, destinationInputs);
+
+    final AuthenticatedUserRequest userRequest =
+        flightContext
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    final BigQueryCow bigQueryCow = crlService.createBigQueryCow(userRequest);
+    // TODO:  remove usage of this client when it's all in CRL
+    final Bigquery bigQueryClient = crlService.createNakedBigQueryClient(userRequest);
+    try {
+      // Get a list of all tables in the source dataset
+      final TableList sourceTables =
+          bigQueryCow
+              .tables()
+              .list(sourceInputs.getProjectId(), sourceInputs.getDatasetName())
+              .execute();
+      // Start a copy job for each source table
+      final Map<String, String> tableToJobId =
+          Optional.ofNullable(
+                  workingMap.get(
+                      ControlledResourceKeys.TABLE_TO_JOB_ID_MAP,
+                      new TypeReference<Map<String, String>>() {}))
```

Comment:
I'm not sure I understand what this is doing. Is this attempting to read an existing map stored in the `workingMap`? If so, when is that populated? AFAIK working map values are only persisted to disk between steps, so this will always be the default value

Reply:
Yes, indeed. I should have caught that! :(

One way to handle it would be to build the map in a separate step of the flight.
Then have a step that iterates through the map starting the jobs. The step would start one job, update the map, and return RERUN status to Stairway. Stairway will persist the flight state and then rerun the same step. The step would find the first, not-yet-started job and start it. We return RERUN until there are no more jobs to start. Then we return SUCCESS and move to the waiting for completion step. 

Another way would be to make a table in the WSM database to hold the state.

In all cases, I think there is a small window where a failure would cause a duplicate copy job to be submitted.

Reply:
What values go in the map before the jobs are started? I don't have a way to create a copy job without starting it. I could make a list of tables in an earlier step if that's helpful.

Reply:
I think you could make a List of a POJO holding `(String table name, @Nullable String jobId)` where all the job ids would start null.

Each time you entered the copy step, you would search the list to find the first null job id. You would start that job, copy the job id into the map and return RERUN. The step boundary would persist the list. Stairway would re-run the step.

There is a window where the job is started, but the list is not persisted where we could still get duplicate jobs. There is simply no way to start the copy job and instantly persist the id. (That is why I prefer the client-provides-the-id style)

Reply:
If Stairway supported something like `flushWorkingMap()` so that I could force an early persistence of it during the step, I wouldn't need to restructure things to add another step and cut this one up. Would that be something worth adding?

Reply:
Sounds good. Implemented using rerun.